src/avx512bw/avx512bw-from-utf8.inl.cpp

```diff
@@ -0,0 +1,466 @@
+                true,   # 0b0111
+                true,   # 0b1000
+                true,   # 0b1001
+                true,   # 0b1010
+                true,   # 0b1011
+                true,   # 0b1100
+                notM,   # 0b1101
+                true,   # 0b1110
+                true,   # 0b1111
+        ] */
+        const __m512i fixup_lookup = _mm512_setr_epi64(
+            0xc0c0c0c0c0c0c000,
+            0xc0c020c0c0c0c0c0,
+            0xc0c0c0c0c0c0c000,
+            0xc0c020c0c0c0c0c0,
+            0xc0c0c0c0c0c0c000,
+            0xc0c020c0c0c0c0c0,
+            0xc0c0c0c0c0c0c000,
+            0xc0c020c0c0c0c0c0
+        );
+
+        const __m512i fixup = _mm512_shuffle_epi8(fixup_lookup, nibble0);
+        const __m512i t0 = _mm512_xor_si512(fixup, continuation1);
+        valid_3bytes = _mm512_mask_cmpgt_epu8_mask(_3bytes, t0, v_1f);
+    }
+
+    // 5. Handle 4-byte chars
+    __mmask64 valid_4bytes;
+    {
+        // continuation1 in range [0..63] (0b0000_0000 .. 0b0011_1111)
+        //
+        // case 1: c1 <= 0xf:  0b0000_xxxx - 0x10 = 0b1111_yyyy
+        // case 2: c1  > 0xf:  0b00xx_xxxx - 0x10 = 0b00yy_yyyy
+        __m512i t0;
+        t0 = _mm512_sub_epi8(continuation1, v_10);
+        t0 = _mm512_xor_si512(t0, v_40);
+        // bit 7th of c: continuation1 <= 0x0f
+        // bit 6th of c: continuation1 >  0x0f
+
+        /** pshufb
+            le_0f   = 0x80  # c1[7] = continuation1 <= 0xf0
+            gt_0f   = 0x40  # c1[6] = not c1[7]
+            true    = gt_0f | le_0f
+            false   = 0x00
+
+            mask_lookup = 4 * [
+                gt_0f,  # 0b0000
+                true,   # 0b0001
+                true,   # 0b0010
+                true,   # 0b0011
+                le_0f,  # 0b0100
+                false,  # 0b0101
+                false,  # 0b0110
+                false,  # 0b0111
+                false,  # 0b1000
+                false,  # 0b1001
+                false,  # 0b1010
+                false,  # 0b1011
+                false,  # 0b1100
+                false,  # 0b1101
+                false,  # 0b1110
+                false,  # 0b1111
+        ] */
+        const __m512i mask_lookup = _mm512_setr_epi64(
+            0x00000080c0c0c040,
+            0x0000000000000000,
+            0x00000080c0c0c040,
+            0x0000000000000000,
+            0x00000080c0c0c040,
+            0x0000000000000000,
+            0x00000080c0c0c040,
+            0x0000000000000000
+        );
+        const __m512i mask = _mm512_shuffle_epi8(mask_lookup, nibble0);
+
+        valid_4bytes = _mm512_mask_test_epi8_mask(_4bytes, mask, t0);
+    }
+
+    {
+        // all: marks all valid leading bytes
+        __mmask64 all = _kor_mask64(valid_2bytes, _kor_mask64(valid_3bytes, valid_4bytes));
+
+        // reset leading byte marks from input
+        valid = _kxor_mask64(valid, all);
+        valid = _kand_mask64(valid, tested_chars);
+    }
+
+    return valid == 0;
```

Comment:
Definitively yes! GCC is really bad and intermixing mask with values. I used to think that it will be smart enough to keep values in mask registers and map basic bit-ops to mask instructions.